File: AVL/avl.c
// An AVL implementation : insertion ONLY

// An AVL tree is balanced either by using,
//  1. Right- Rotation
//  2. Left- Rotation


// ??????????????????????????????????????????????
// ???                                        ???
// ???    Adelson-Velskii and Landis : AVL    ???
// ???                                        ???
// ??????????????????????????????????????????????

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

// struct node 
struct node{

   int key;
   struct node *left;
   struct node *right;
   int height;
};

typedef struct node NODE;

// function declarations
NODE *leftrotate(NODE *);
NODE *rightrotate(NODE *);
NODE *insert(NODE *,int );
NODE *newNode(int);
NODE *inorder(NODE*);
int max(int, int);
int height(NODE *);


int main(){

    NODE *root = NULL;
    int arr[] ={34, 78, 90, 45, 67, 89, 12, 342}, i;
    root = insert(root , 40);
    for(i=0; i<sizeof(arr)/sizeof(arr[0]); i++){
        
        insert(root, arr[i]);
    }

    inorder(root);
    printf("\n");

	return 0;
}

// utility to add new node maintaining the searching property.
NODE *insert(NODE *root, int key){

	if(root == NULL)
		return newNode(key);
	else if(key < root->key)
		root->left = insert(root->left, key);
	else if(key > root->key)
		root->right = insert(root->right, key);
	else{
		printf("equal nodes are not allowed\n");
	    return root;
    }
   
    root->height = 1 + max(height(root->left), height(root->right));
    
    int balance = get_bal(root);
    // there are four cases if this node is unbalanced.
  

    //left left case
    if(balance > 1 && key < root->left->key)
    	return rightrotate(root);
    // right right case
    if(balance < -1 && key > root->right->key)
    	return leftrotate(root);
    //left right case
    if(balance > 1 && key > root->left->key){
    	root->left = leftrotate(root->left);
    	return rightrotate(root);
    }
    // right left case
    if(balance < -1 && key < root->right->key){
    	root->right = rightrotate(root->right);
    	return leftrotate(root);
    }
    




	return root;
}

NODE *leftrotate(NODE *x){

	NODE *y = x->right;
	NODE *T2 = y->left;

    // perform rotation
	y->left = x;
	x->right = T2;

    // update heights
	x->height = max(height(x->left), height(x->right)) + 1;
	y->height = max(height(y->left), height(y->right))+ 1;

    // return new root;
	return y;

}

NODE *rightrotate(NODE *y){

	NODE *x = y->left;
	NODE *T2 = x->right;

    // perform rotation
	x->right = y;
	y->left = T2;

    // update heights
	y->height = max(height(y->left), height(y->right)) + 1;
	x->height = max(height(x->left), height(x->right)) + 1;

    // return new root
	return x;
}

// helper utility that returns new node with key = key and height initially one.
NODE *newNode(int key){

	NODE *node = (NODE *)malloc(sizeof(NODE));
	node->key = key;
	node->height = 1; // height is initially one.
	node->left = node->right = NULL;

	return node;
}


// utility that helps in printing the tree in inorder or sorted order(increasing).
NODE *inorder(NODE *node){

	if(node != NULL){

		inorder(node->left);
		printf("%d\t", node->key);
		inorder(node->right);
	}

}




//!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
//!!!!!!!!!!!!!!!   HELPER FUNCTIONS  !!!!!!!!!!!!!!!!!!!


// helper function that returns the max of two
int max(int a, int b){

	return (a> b)? a : b;
}

// utility that returns the height of the node
int height(NODE *node){

      if (node == NULL)
      	return 0;
      else
      	return node->height;
}


// utility that return the balance factor
int get_bal(NODE *node){

	if(node == NULL)
		return 0;
	else 
		return height(node->left)- height(node->right);
}

File: python/avl and red black/avl.py
"""avl in python: insertion only"""


	
class Node:
	def __init__(self, value):
		self.data = value
		self.r_child = None
		self.l_child = None
		self.height = 0


def insert(root, node):
	if root is None:
		root = node
	else:
		if node.data < root.data:
			if root.l_child is None:
				root.l_child = node
			else:
				insert(root.l_child, node)
		elif node.data > root.data:
			if root.r_child is None:
				root.r_child = node
			else:
				insert(root.r_child, node)
		else:
			print "error\n",root.data

		node.height = 1+ max(height(node.l_child), node.r_child)
		balance = get_bal(node)
		# four conditions LL, LR, RR, RL

		if balance > 1: #LL or LR
			if height(node.l_child.l_child) > height(node.l_child.r_child): #LL
				return rotateright(node)
			else:
				node.l_child = rotateleft(node)
				return  rotateright(node)

		elif balance < -1:
			if height(node.r_child.r_child) > height(node.r_child.l_child): #RR
				return rotateleft(node)
			else:
				node.r_child = rotateright(node)
				return rotateleft(node)


def inorder(root):
	if root is not None:
		inorder(root.l_child)
		print root.data
		inorder(root.r_child)

def rotateright(node):
	new = node.r_child
	new2 = new.l_child

	new.l_child = node
	node.r_child = new2

	node.height =max(height(node.l_child), height(node.r_child)) +1
	new.height = max(height(new.l_child), height(new.r_child)) + 1

	return new

def rotateleft(node):
	new = node.l_child
	new2 = new.r_child

	new.r_child = node
	new.l_child = new2

	node.height = max(height(node.l_child), height(node.r_child)) + 1
	new.height = max(height(new.l_child), height(new.r_child)) + 1

	return new

def get_bal(node):
	if node is None:
		return 0
	else:
		return height(node.l_child) - height(node.r_child)

def height(node):
	if node is None:
		return 0
	else:
		return node.height

def max(a, b):
	if a > b:
		return a
	else:
		return b


r = Node(900)
for i in [1,98,45,23,9,0,12]:
	insert(r, Node(i))

inorder(r)